### PR TITLE
documentation building, readthedocs config

### PR DESCRIPTION
**Pull request by newville**
this is a work-in-progress to start on getting docs to build and push to readthedocs. 

so far, it has not touched the doc contents, but some of that will need changing as well.

### CODE REVIEW

**Comment by newville** [comment]
This PR to update the docs is now at the state of being incomplete but reviewable. 

By way of documentation for what has changed in this PR:

1.  This adds config files for codecov and readthedocs.   These might need more tweaking or have been updated already in earlier PRs.  
2. some sphinx changes were needed to the layout. 
3. I made "index" a lot shorter. It is now mostly an intro and Table of Contents. I added a custom "sidebar" giving the version, quick install instructions, and a link to GitHub. 
4. moved most of what was in `index.rst` to `install.rst`, but heavily edited this to more modern install instructions, and updated Credits to describe the recent transition to the lmfit github org.
5. made some edits to "user_guide",  but more could be done.  No substantial changes to "numpy" or "tech guide".
6. I updated the docstrings for `ufloat` and `ufloat_fromstr`, so that "autodoc" can be (and is) used with sphinx.
7. for linking to PDF and static zipped HTML, the current approach is fragile, so I sort of set these up differently:

I am fine with using readthedocs.  But, I am also willing to use GitHub pages.  I use this for almost all my other projects, and it has the distinct advantage of allowing us to put any files we want in the gh-pages branch, including static PDFs.   

By way of proof-of-concept and viewing the development here, I added a gh-pages branch to the main `lmfit/uncertainties` and pushed the rendered results of the current state of this PR there.   This is viewable at https://lmfit.github.io/uncertainties/

We could use both, and/or make the links to the PDF and zipped HTML point to the fully-qualified URLs https://lmfit.github.io/uncertainties/uncertainties.pdf and https://lmfit.github.io/uncertainties/uncertainties_doc.zip.

I think there is always more to do.   And I would say that all the changes here could be undone or modified: suggestions most welcome.

File: doc/user_guide.rst [hunk 1]
```diff
@@ -6,69 +6,107 @@ User Guide
...
-The :func:`ufloat` function creates numbers with uncertainties. Existing
-calculation code can usually run with no or little modification and
-automatically produce results with uncertainties.
+The :mod:`uncertainties` module contains sub-modules for :ref:`advanced
+mathematical functions <advanced math operations>`, and :doc:`arrays and
+matrices <numpy_guide>`, which can be accessed with
 
-.. The "import uncertainties" is put here because some examples requires
-   uncertainties to have been imported (and not only ufloat).
+>>> import uncertainties as un
```
Comment (andrewgsavage) [inline doc/user_guide.rst:21]: does this conflict with any other packages?

Comment (newville) [reply doc/user_guide.rst:21]: dunno.  One could probably ask that about any abbreviation.

**Comment by newville** [comment]
@andrewgsavage 
> the table of contents has switched position with 'get uncertainties' on index.html

Yes.  The intention is that the information on how to get uncertainties is prominent on the index page. 
Is this a problem? 

> I'd remove the autofuncton then this looks good to merge

Do you mean you want to not have the docstrings for `ufloat` and `ufloat_fromstr`?  One of the (okay, many) things I find odd about the current doc is that these are not provided. And the docstrings are kind of a mess.  

I think these autofunctions could be moved, but I think they should be given somewhere.

**Comment by andrewgsavage** [comment]
That's fine, the 'get uncertainties' on the box on the right where the toc
would be looks odd


I mean to move them to another page. Personally I'd leave that page till a
later PR when there's more updated doc strings, but you could do so in this
PR.

File: uncertainties/core.py [hunk 2]
```diff
@@ -2692,12 +2693,10 @@ class NegativeStdDev(Exception):
...
 class Variable(AffineScalarFunc):
     """
-    Representation of a float-like scalar random variable, along with
-    its uncertainty.
+    Representation of a float-like scalar Variable with its uncertainty.
 
-    Objects are meant to represent variables that are independent from
-    each other (correlations are handled through the AffineScalarFunc
-    class).
+    Variablees are independent from each other, but correlations between them
```
Comment (andrewgsavage) [inline uncertainties/core.py:2698]: Variablees >Variables

**Comment by andrewgsavage** [comment]
Looks good, how much more are you planning to do on this before merging? this + changelog are the last things to do before a new release.

File: doc/user_guide.rst [hunk 3]
```diff
@@ -6,69 +6,107 @@ User Guide
...
+   >>> x = ufloat_fromstr("0.20+/-0.01")
+   >>> x = ufloat_fromstr("(2+/-0.1)e-01")  # Factored exponent
+   >>> x = ufloat_fromstr("0.20(1)")  # Short-hand notation
+   >>> x = ufloat_fromstr("20(1)e-2")  # Exponent notation
+   >>> x = ufloat_fromstr(u"0.20±0.01")  # Pretty-print form
+   >>> x = ufloat_fromstr("0.20")  # Automatic uncertainty of +/-1 on last digit
 
-Each number created this way is an **independent (random) variable**
-(for details, see the :ref:`Technical Guide <math_def_num_uncert>`).
+.. autofunction:: ufloat
```
[the remark this answers is not in the document]
Comment (newville) [reply doc/user_guide.rst:73]: That seems OK.   The current docs don't describe Variable, UFloat, or AffineScalarFunc. They sort of talk about "value with an uncertainty".   I think that an API section really should mention this.    Opinions?

File: doc/formatting.rst [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+.. Formatting method:
+
+More **control over the format** can be obtained (in Python 2.6+)
+through the usual :func:`format` method of strings:
+
+>>> print('Result = {:10.2f}'.format(x))
+Result =       0.20+/-      0.01
+
+(Python 2.6 requires ``'{0:10.2f}'`` instead, with the usual explicit
```
[the remark this answers is not in the document]
Comment (newville) [reply doc/formatting.rst:41]: hah! thanks - I missed that one.

File: doc/formatting.rst [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+Data Group
+<http://PDG.lbl.gov/2010/reviews/rpp2010-rev-rpp-intro.pdf>`_ rounding
+rules (these rules keep the number of digits small, which is
+convenient for reading numbers with uncertainties, and at the same
+time prevent the uncertainty from being displayed with too few
+digits):
+
+>>> print('Automatic number of digits on the uncertainty: {}'.format(x))
+Automatic number of digits on the uncertainty: 0.200+/-0.010
+>>> print(x0
```
[the remark this answers is not in the document]
Comment (newville) [reply doc/formatting.rst:88]: +1

File: doc/user_guide.rst [hunk 6]
```diff
@@ -6,130 +6,273 @@ User Guide
...
-The :func:`ufloat` function creates numbers with uncertainties. Existing
-calculation code can usually run with no or little modification and
-automatically produce results with uncertainties.
+The :mod:`uncertainties` module contains sub-modules for :ref:`advanced
+mathematical functions <advanced math operations>`, and :doc:`arrays and
+matrices <numpy_guide>`, which can be accessed with::
 
-.. The "import uncertainties" is put here because some examples requires
-   uncertainties to have been imported (and not only ufloat).
+ >>> import uncertainties
```
[the remark this answers is not in the document]
Comment (newville) [reply doc/user_guide.rst:23]: Yep, I agree...

File: doc/user_guide.rst [hunk 7]
```diff
@@ -6,130 +6,273 @@ User Guide
...
+.. autofunction:: ufloat
+
+.. autofunction:: ufloat_fromstr
+
+.. autoclass:: Variable
+
+Basic math with uncertain Variables
+=========================================
+
+:class:`Variable`s can be used in basic mathematical calculations
```
[the remark this answers is not in the document]
Comment (newville) [reply doc/user_guide.rst:82]: will investigate (when moved to an `api.rst`, probably)

File: doc/user_guide.rst [hunk 8]
```diff
@@ -6,130 +6,273 @@ User Guide
...
-=======================
+Mathematical operations with uncertain Variables
+=====================================================
 
-Besides being able to apply basic mathematical operations to numbers
-with uncertainty, this package provides generalizations of **most of
-the functions from the standard** :mod:`math` **module**.  These
-mathematical functions are found in the :mod:`uncertainties.umath`
-module:
+Besides being able to apply basic mathematical operations to uncertaintties
```
[the remark this answers is not in the document]
Comment (newville) [reply doc/user_guide.rst:150]: +1

File: doc/user_guide.rst [hunk 9]
```diff
@@ -6,130 +6,273 @@ User Guide
...
+>>> (x+y).std_dev == sqrt(x.std_dev**2 + y.std_dev**2)
+True
+
+
+Multiplying two Variables will properly propagate those
+uncertainties too:
+
+>>> print(x*y)
+240.0+/-76.83749084919418
+>>> (x*y).std_dev == (x*y).nominal_value *  sqrt((x.std_dev/x.nominal_value)**2 + (y.std_dev/y.nominal_value)**2 )
```
[the remark this answers is not in the document]
Comment (newville) [reply doc/user_guide.rst:113]: OK, I was thinking maybe avoid `.n` and `.s` as they might be something to deprecate.  I'm OK with fully embracing them as part of the API.

File: uncertainties/core.py (hunk 2 above)
Comment (newville) [reply uncertainties/core.py:2698]: +1

**Comment by newville** [comment]
Yeah, how much more to do is a great question ;).    

I might ask opinions all around (@wshanks, @jagerber48) on how complete this needs to be, and if the version at https://lmfit.github.io/uncertainties is approaching "acceptable".

**Comment by newville** [comment]
OK, sorry for the delay on this. I believe I have addressed most of the issues and this might be good enough, at least for a while.   Suggestions?  Again, rendered results are at https://lmfit.github.io/uncertainties

File: doc/install.rst [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+Credits
+================
+
+.. _Eric O. LEBIGOT (EOL): http://linkedin.com/pub/eric-lebigot/22/293/277
+
+The :mod:`uncertainties` package was written and developed by `Eric O. LEBIGOT
+(EOL)`_.  EOL also maintained the package until 2024, when the GitHub project
+was moved to the `lmfit GitHub organization`_ to allow more sustainable
+development and maintenance.  Currentt members of the devlopment and
```
Comment (andrewgsavage) [inline doc/install.rst:106]: Current

File: doc/user_guide.rst [hunk 11]
```diff
@@ -6,130 +6,267 @@ User Guide
...
+   pair: number with uncertainty; creation
 
-The :mod:`uncertainties` module contains other features, which can be
-made accessible through
+Creating Variables: numbers with uncertainties
+================================================
 
->>> import uncertainties
+To create a number with uncertainties or *Variable*, use the :func:`ufloat`
+function, which takes a *nominal value* ( (which can be interpreted as the most
```
Comment (andrewgsavage) [inline doc/user_guide.rst:30]: Double open bracket

File: doc/user_guide.rst [hunk 12]
```diff
@@ -6,130 +6,267 @@ User Guide
...
+likely value, or the mean or central value of the distribution of values), a
+*standard error* ( (the standard deviation or :math:`1-\sigma` uncertainty), and
```
Comment (andrewgsavage) [inline doc/user_guide.rst:32]: Double open bracket

File: doc/user_guide.rst [hunk 13]
```diff
@@ -6,130 +6,267 @@ User Guide
...
+
+
+Comparisons of magnitude
+------------------------------------
+
+The concept of comparing the magnitude of values with uncertainties is a bit
+complicated.  That is, a Variable with a value of 25 +/- 10 might be greater
+than a Variable with a value of 24 +/- 8 most of the time, but *sometimes* it
+might be less than it.   The :mod:`uncertainties` package takes the simple
+approach of comparing.  That is
```
Comment (andrewgsavage) [inline doc/user_guide.rst:223]: Of comparing nominal values.

**Comment by andrewgsavage** [comment]
There's some python 2 print statements in the tech guide

**Comment by andrewgsavage** [comment]
Also some python 2 print statements on the bumpy guide

**Comment by andrewgsavage** [comment]
Also some python 2 print statements on the numpy guide

**Comment by andrewgsavage** [comment]
Numpy*
There's also a ref to python2 on the numpy page.

For Python 2:

converters = dict.fromkeys(range(num_cols), uncertainties.ufloat_fromstr)



Other than that this looks and reads well. Id fix those then merge.

**Comment by newville** [comment]
@andrewgsavage Thanks!! OK, I think I have addressed hose spelling/print errors. 

I am still investigating the "loadtxt" thing:  "The code does what???"

**Comment by newville** [comment]
@andrewgsavage OK, I think the loadtxt/savetxt doc is correct now.

**Review by andrewgsavage — APPROVED** [review]

**Comment by newville** [comment]
@andrewgsavage Thanks!  Should we re-kindle the conversation about tagging and releasing 3.2.0?

**Comment by andrewgsavage** [comment]
Looks good, merged since there's been no indication anyone else will read through since you asked 2 weeks ago.

**Comment by andrewgsavage** [comment]
> @andrewgsavage Thanks! Should we re-kindle the conversation about tagging and releasing 3.2.0?

yea please

**Comment by jagerber48** [comment]
Why did we remove `make.bat`? My normal workflow uses that to build docs locally on my Windows machine. Is there another workflow to build docs locally?

**Comment by newville** [comment]
@jagerber48 I think we removed 'make.bat' because we were not actively using it and did not want to keep something that we weren't using/testing/maintaining.     It would be OK with me to add that back. 

Using `sphinx-quickstart` would probably give a decent `make.bat` file to start with.

**Pull request merged**